<a href="https://colab.research.google.com/github/nopy4ik/AirbnbDataAnalysisHW/blob/main/airbnb_homework.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [51]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime

airbnb_data = pd.read_csv('/content/Airbnb_Open_Data.csv')


airbnb_data.head(3)

/tmp/ipython-input-1133937323.py:7: DtypeWarning: Columns (25) have mixed types. Specify dtype option on import or set low_memory=False.
  airbnb_data = pd.read_csv('/content/Airbnb_Open_Data.csv')


id                                 NAME      host id  \
0  1001254   Clean & quiet apt home by the park  80014485718   
1  1002102                Skylit Midtown Castle  52335172823   
2  1002403  THE VILLAGE OF HARLEM....NEW YORK !  78829239556   

  host_identity_verified host name neighbourhood group neighbourhood  \
0            unconfirmed  Madaline            Brooklyn    Kensington   
1               verified     Jenna           Manhattan       Midtown   
2                    NaN     Elise           Manhattan        Harlem   

        lat      long        country  ... service fee minimum nights  \
0  40.64749 -73.97237  United States  ...       $193            10.0   
1  40.75362 -73.98377  United States  ...        $28            30.0   
2  40.80902 -73.94190  United States  ...       $124             3.0   

  number of reviews last review  reviews per month review rate number  \
0               9.0  10/19/2021               0.21                4.0   
1              45.0   5/21/2022               0.38                4.0   
2               0.0         NaN                NaN                5.0   

  calculated host listings count  availability 365  \
0                            6.0             286.0   
1                            2.0             228.0   
2                            1.0             352.0   

                                         house_rules license  
0  Clean up and treat the home the way you'd like...     NaN  
1  Pet friendly but please confirm with me if the...     NaN  
2  I encourage you to use my kitchen, cooking and...     NaN  

[3 rows x 26 columns]

In [7]:
airbnb_data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102599 entries, 0 to 102598
Data columns (total 26 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   id                              102599 non-null  int64  
 1   NAME                            102349 non-null  object 
 2   host id                         102599 non-null  int64  
 3   host_identity_verified          102310 non-null  object 
 4   host name                       102193 non-null  object 
 5   neighbourhood group             102570 non-null  object 
 6   neighbourhood                   102583 non-null  object 
 7   lat                             102591 non-null  float64
 8   long                            102591 non-null  float64
 9   country                         102067 non-null  object 
 10  country code                    102468 non-null  object 
 11  instant_bookable                102494 non-null  object 
 12  cancellation_pol

In [8]:
airbnb_data.describe()


id       host id            lat           long  \
count  1.025990e+05  1.025990e+05  102591.000000  102591.000000   
mean   2.914623e+07  4.925411e+10      40.728094     -73.949644   
std    1.625751e+07  2.853900e+10       0.055857       0.049521   
min    1.001254e+06  1.236005e+08      40.499790     -74.249840   
25%    1.508581e+07  2.458333e+10      40.688740     -73.982580   
50%    2.913660e+07  4.911774e+10      40.722290     -73.954440   
75%    4.320120e+07  7.399650e+10      40.762760     -73.932350   
max    5.736742e+07  9.876313e+10      40.916970     -73.705220   

       Construction year  minimum nights  number of reviews  \
count      102385.000000   102190.000000      102416.000000   
mean         2012.487464        8.135845          27.483743   
std             5.765556       30.553781          49.508954   
min          2003.000000    -1223.000000           0.000000   
25%          2007.000000        2.000000           1.000000   
50%          2012.000000        3.000000           7.000000   
75%          2017.000000        5.000000          30.000000   
max          2022.000000     5645.000000        1024.000000   

       reviews per month  review rate number  calculated host listings count  \
count       86720.000000       102273.000000                   102280.000000   
mean            1.374022            3.279106                        7.936605   
std             1.746621            1.284657                       32.218780   
min             0.010000            1.000000                        1.000000   
25%             0.220000            2.000000                        1.000000   
50%             0.740000            3.000000                        1.000000   
75%             2.000000            4.000000                        2.000000   
max            90.000000            5.000000                      332.000000   

       availability 365  
count     102151.000000  
mean         141.133254  
std          135.435024  
min          -10.000000  
25%            3.000000  
50%           96.000000  
75%          269.000000  
max         3677.000000

In [9]:

airbnb_data.isnull().sum()

id                                     0
NAME                                 250
host id                                0
host_identity_verified               289
host name                            406
neighbourhood group                   29
neighbourhood                         16
lat                                    8
long                                   8
country                              532
country code                         131
instant_bookable                     105
cancellation_policy                   76
room type                              0
Construction year                    214
price                                247
service fee                          273
minimum nights                       409
number of reviews                    183
last review                        15893
reviews per month                  15879
review rate number                   326
calculated host listings count       319
availability 365                     448
house_rules                        52131
license                           102597
dtype: int64

Вывод: в датасете 102599 наблюдений и 26 признаков. Присутствуют пропущенные значения. Также есть несоответствия в типах данных (например, год постройки и дата последнего отзыва)

Детали, которые заметили:

Объявления содержат в себе жилье с датой постройки с 2003 по 2022 год включительно
Минимальное кол-во ночей -1223 (ошибка?) и аномально высокое максимальное значение (заполнять пропуски надо медианой, а не средним)
Минимальное значение доступности 365 равно -10 (ошибка?) и очень высокое максимальное значение (заполнять пропуски надо медианой, а не средним)
Колонки страна и код страны содержат только 1 уникальное значение
Столбец license имеет только 2 заполненных значения
Столбцы host_identity_verified и instant_bookable бинарные (имеют только 2 уникальных значения)
В house_rules наиболее часто встречающееся значение '#NAME?' похоже на ошибку считывания данных или пропуск

In [10]:
copy_data = airbnb_data.copy()

In [11]:
copy_data.columns = [c.replace(" ", "_").lower() for c in copy_data.columns]
copy_data.columns

Index(['id', 'name', 'host_id', 'host_identity_verified', 'host_name',
       'neighbourhood_group', 'neighbourhood', 'lat', 'long', 'country',
       'country_code', 'instant_bookable', 'cancellation_policy', 'room_type',
       'construction_year', 'price', 'service_fee', 'minimum_nights',
       'number_of_reviews', 'last_review', 'reviews_per_month',
       'review_rate_number', 'calculated_host_listings_count',
       'availability_365', 'house_rules', 'license'],
      dtype='object')

In [12]:
pd.DataFrame(copy_data.isna().mean()).sort_values(by=0, ascending=False).style.format('{:.2%}').background_gradient('coolwarm')

- Столбец license может содержать информацию о юридических требованиях или лицензиях, необходимых для эксплуатации объявления на Airbnb. Возможно, что не все хозяева предоставляют информацию о лицензии, что приводит к отсутствию значений в этом столбце. Или некоторым хозяевам может не потребоваться предоставлять информацию о лицензии в зависимости от местных правил или типа объявления. Или это нововведение и такая информация ещё не успела накопиться.

- Пропуски в столбце house_rules могут говорить о нежелании хозяев заполнять данную информацию или о том, что у хозяев нет никаких дополнений к стандартным правилам аренды жилья.
- Отсутствующие значения в last_review и reviews_per_month могут указывать на объявления, которые никогда не получали отзывов, то есть они только что были выставлены на площадку или гости еще не успели оставить.

In [13]:
copy_data.columns

Index(['id', 'name', 'host_id', 'host_identity_verified', 'host_name',
       'neighbourhood_group', 'neighbourhood', 'lat', 'long', 'country',
       'country_code', 'instant_bookable', 'cancellation_policy', 'room_type',
       'construction_year', 'price', 'service_fee', 'minimum_nights',
       'number_of_reviews', 'last_review', 'reviews_per_month',
       'review_rate_number', 'calculated_host_listings_count',
       'availability_365', 'house_rules', 'license'],
      dtype='object')

In [14]:
print(f'Размер датасета до изменения: {copy_data.shape}')
copy_data = copy_data.drop(['license', 'country', 'country_code'], axis=1)
print(f'Размер датасета после изменения: {copy_data.shape}')

Размер датасета до изменения: (102599, 26)
Размер датасета после изменения: (102599, 23)


In [15]:
copy_data.house_rules.unique()[:5]

array(["Clean up and treat the home the way you'd like your home to be treated.  No smoking.",
       'Pet friendly but please confirm with me if the pet you are planning on bringing with you is OK. I have a cute and quiet mixed chihuahua. I could accept more guests (for an extra fee) but this also needs to be confirmed beforehand. Also friends traveling together could sleep in separate beds for an extra fee (the second bed is either a sofa bed or inflatable bed). Smoking is only allowed on the porch.',
       "I encourage you to use my kitchen, cooking and laundry facilities. There is no additional charge to use the washer/dryer in the basement.  No smoking, inside or outside. Come home as late as you want.  If you come home stumbling drunk, it's OK the first time. If you do it again, and you wake up me or the neighbors downstairs, we will be annoyed.  (Just so you know . . . )",
       nan,
       'Please no smoking in the house, porch or on the property (you can go to the nearby cor

In [16]:
print(f'Размер датасета до изменения: {copy_data.shape}')
copy_data['house_rules'].fillna("No house rules provided", inplace=True)
print(f'Размер датасета после изменения: {copy_data.shape}')

Размер датасета до изменения: (102599, 23)
Размер датасета после изменения: (102599, 23)


/tmp/ipython-input-62481278.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  copy_data['house_rules'].fillna("No house rules provided", inplace=True)


In [17]:
print(f'Размер датасета до изменения: {copy_data.shape}')
nan_columns = ['last_review', 'reviews_per_month', 'number_of_reviews', 'review_rate_number']
for col in nan_columns:
    copy_data[col].fillna(0, inplace=True)
print(f'Размер датасета после изменения: {copy_data.shape}')

Размер датасета до изменения: (102599, 23)
Размер датасета после изменения: (102599, 23)


/tmp/ipython-input-3256924158.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  copy_data[col].fillna(0, inplace=True)


# instant_bookable, cancellation_policy, host_identity_verified

In [18]:
print(f'Размер датасета до изменения: {copy_data.shape}')
categorical_columns = ['instant_bookable', 'cancellation_policy']
for col in categorical_columns:
    copy_data[col].fillna(copy_data[col].mode()[0], inplace=True)
print(f'Размер датасета после изменения: {copy_data.shape}')

Размер датасета до изменения: (102599, 23)
Размер датасета после изменения: (102599, 23)


/tmp/ipython-input-2406554971.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  copy_data[col].fillna(copy_data[col].mode()[0], inplace=True)
/tmp/ipython-input-2406554971.py:4: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  copy_data[col].fillna(copy_data[col].mode()[0], inplace=True)


In [19]:
# заполним пропущенные значения в столбце host_identity_verified значение unconfirmed
copy_data['host_identity_verified'].fillna('unconfirmed', inplace=True)

/tmp/ipython-input-3985934960.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  copy_data['host_identity_verified'].fillna('unconfirmed', inplace=True)


In [20]:
copy_data.isna().sum()

id                                  0
name                              250
host_id                             0
host_identity_verified              0
host_name                         406
neighbourhood_group                29
neighbourhood                      16
lat                                 8
long                                8
instant_bookable                    0
cancellation_policy                 0
room_type                           0
construction_year                 214
price                             247
service_fee                       273
minimum_nights                    409
number_of_reviews                   0
last_review                         0
reviews_per_month                   0
review_rate_number                  0
calculated_host_listings_count    319
availability_365                  448
house_rules                         0
dtype: int64

# neighbourhood_group, neighbourhood, lat, long

In [21]:
# Выведем кол-во строк, в которых пропущены эти значения
copy_data[copy_data[['neighbourhood_group', 'neighbourhood', 'lat', 'long']].isna().any(axis=1)].shape[0]

53

In [22]:
print(f'Размер датасета до изменения: {copy_data.shape}')
copy_data = copy_data.dropna(subset=['neighbourhood_group', 'neighbourhood', 'lat', 'long'])
print(f'Размер датасета после изменения: {copy_data.shape}')

Размер датасета до изменения: (102599, 23)
Размер датасета после изменения: (102546, 23)


# name, host_name

In [23]:
print(f'Размер датасета до изменения: {copy_data.shape}')
copy_data['name'].fillna("unknown", inplace=True)
copy_data['host_name'].fillna("unknown", inplace=True)
print(f'Размер датасета после изменения: {copy_data.shape}')

Размер датасета до изменения: (102546, 23)
Размер датасета после изменения: (102546, 23)


/tmp/ipython-input-3608613252.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  copy_data['name'].fillna("unknown", inplace=True)
/tmp/ipython-input-3608613252.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try u

In [24]:
copy_data.dtypes

id                                  int64
name                               object
host_id                             int64
host_identity_verified             object
host_name                          object
neighbourhood_group                object
neighbourhood                      object
lat                               float64
long                              float64
instant_bookable                     bool
cancellation_policy                object
room_type                          object
construction_year                 float64
price                              object
service_fee                        object
minimum_nights                    float64
number_of_reviews                 float64
last_review                        object
reviews_per_month                 float64
review_rate_number                float64
calculated_host_listings_count    float64
availability_365                  float64
house_rules                        object
dtype: object

# price, service_fee

In [25]:
copy_data.loc[0, 'price']

'$966 '

In [26]:
copy_data.loc[0, 'service_fee']

'$193 '

In [27]:
def replace_symbols(string):
    try:
        return float(string.replace('$', '').replace(',', '').strip())
    except:
        return string

In [28]:
copy_data['service_fee'] = copy_data['service_fee'].apply(replace_symbols)
copy_data['price'] = copy_data['price'].apply(replace_symbols)

In [29]:
copy_data['service_fee'].head()

0    193.0
1     28.0
2    124.0
3     74.0
4     41.0
Name: service_fee, dtype: float64

In [30]:
copy_data['price'].head()

0    966.0
1    142.0
2    620.0
3    368.0
4    204.0
Name: price, dtype: float64

In [31]:
# заполняем нулём пропущенные значения в столбце сервисного сбора
copy_data['service_fee'].fillna(0, inplace=True)

/tmp/ipython-input-210584305.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  copy_data['service_fee'].fillna(0, inplace=True)


In [32]:
copy_data['room_type'].unique()

array(['Private room', 'Entire home/apt', 'Shared room', 'Hotel room'],
      dtype=object)

In [33]:
copy_data['neighbourhood_group'].unique()

array(['Brooklyn', 'Manhattan', 'brookln', 'manhatan', 'Queens',
       'Staten Island', 'Bronx'], dtype=object)

In [34]:
# приведем названия всех районов к одному виду
copy_data.loc[copy_data['neighbourhood_group'] == 'brookln', 'neighbourhood_group'] = 'Brooklyn'
copy_data.loc[copy_data['neighbourhood_group'] == 'manhatan', 'neighbourhood_group'] = 'Manhattan'

In [35]:
copy_data.groupby(['room_type', 'neighbourhood_group'])['price'].median()

room_type        neighbourhood_group
Entire home/apt  Bronx                  613.0
                 Brooklyn               626.0
                 Manhattan              622.0
                 Queens                 624.0
                 Staten Island          667.0
Hotel room       Brooklyn               836.0
                 Manhattan              644.5
                 Queens                 367.5
Private room     Bronx                  652.0
                 Brooklyn               624.0
                 Manhattan              619.0
                 Queens                 629.0
                 Staten Island          588.0
Shared room      Bronx                  608.0
                 Brooklyn               662.0
                 Manhattan              640.5
                 Queens                 659.5
                 Staten Island          680.0
Name: price, dtype: float64

In [36]:
copy_data['median_price_by_type_and_group'] = copy_data.groupby(['room_type', 'neighbourhood_group'])['price'].transform('median')
copy_data[['room_type', 'neighbourhood_group', 'median_price_by_type_and_group']]

room_type neighbourhood_group  median_price_by_type_and_group
0          Private room            Brooklyn                           624.0
1       Entire home/apt           Manhattan                           622.0
2          Private room           Manhattan                           619.0
3       Entire home/apt            Brooklyn                           626.0
4       Entire home/apt           Manhattan                           622.0
...                 ...                 ...                             ...
102594     Private room            Brooklyn                           624.0
102595     Private room           Manhattan                           619.0
102596     Private room            Brooklyn                           624.0
102597  Entire home/apt              Queens                           624.0
102598  Entire home/apt           Manhattan                           622.0

[102546 rows x 3 columns]

In [37]:
copy_data['price'] = copy_data['price'].mask(pd.isnull, copy_data['median_price_by_type_and_group'])

In [38]:
copy_data.columns[copy_data.isna().sum() != 0]

Index(['construction_year', 'minimum_nights', 'calculated_host_listings_count',
       'availability_365'],
      dtype='object')

# availability_365, construction_year, minimum_nights, calculated_host_listings_count

In [39]:
# посмотрим на медианные значения для этих столбцов
columns = ['availability_365', 'construction_year', 'minimum_nights', 'calculated_host_listings_count']
for col in columns:
    print(f'{col} - медиана {copy_data[col].median()}')

availability_365 - медиана 96.0
construction_year - медиана 2012.0
minimum_nights - медиана 3.0
calculated_host_listings_count - медиана 1.0


In [40]:
columns = ['availability_365', 'minimum_nights']
for col in columns:
    print(f'{col} - медиана {copy_data.loc[copy_data[col] >= 0, col].median()}')

availability_365 - медиана 98.0
minimum_nights - медиана 3.0


In [41]:
# заполним медианой пропущенные значения
columns = ['availability_365', 'construction_year', 'minimum_nights', 'calculated_host_listings_count']
for col in columns:
    copy_data[col].fillna(copy_data[col].median(), inplace=True)

/tmp/ipython-input-2150648107.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  copy_data[col].fillna(copy_data[col].median(), inplace=True)


In [42]:
copy_data.columns[copy_data.isna().sum() != 0]

Index([], dtype='object')

In [43]:
copy_data.head(2)

id                                name      host_id  \
0  1001254  Clean & quiet apt home by the park  80014485718   
1  1002102               Skylit Midtown Castle  52335172823   

  host_identity_verified host_name neighbourhood_group neighbourhood  \
0            unconfirmed  Madaline            Brooklyn    Kensington   
1               verified     Jenna           Manhattan       Midtown   

        lat      long  instant_bookable  ... service_fee minimum_nights  \
0  40.64749 -73.97237             False  ...       193.0           10.0   
1  40.75362 -73.98377             False  ...        28.0           30.0   

   number_of_reviews  last_review  reviews_per_month  review_rate_number  \
0                9.0   10/19/2021               0.21                 4.0   
1               45.0    5/21/2022               0.38                 4.0   

   calculated_host_listings_count availability_365  \
0                             6.0            286.0   
1                             2.0            228.0   

                                         house_rules  \
0  Clean up and treat the home the way you'd like...   
1  Pet friendly but please confirm with me if the...   

   median_price_by_type_and_group  
0                           624.0  
1                           622.0  

[2 rows x 24 columns]

In [44]:
copy_data['host_identity_verified'].value_counts()

host_identity_verified
unconfirmed    51459
verified       51087
Name: count, dtype: int64

In [45]:
# переводим бинарный признак в признак со значениями True/False
print(f'Размер датасета до изменения: {copy_data.shape}')
copy_data['is_host_identity_verified'] = copy_data['host_identity_verified'] == 'verified' # создаём новый бинарный признак "верифицирован ли хост"
copy_data.drop('host_identity_verified', axis=1, inplace=True) # удаляем уже ненужный признак
print(f'Размер датасета после изменения: {copy_data.shape}')

Размер датасета до изменения: (102546, 24)
Размер датасета после изменения: (102546, 24)


In [46]:
copy_data['instant_bookable'].value_counts()

instant_bookable
False    51549
True     50997
Name: count, dtype: int64

In [47]:
# конвертируем столбец с возможностью мгновенного бронирования в булев тип данных
copy_data['instant_bookable'] = copy_data['instant_bookable'].astype(bool)
copy_data['instant_bookable'].dtype

dtype('bool')

In [48]:
# конвертируем год постройки в целочисленный тип данных
columns = ['construction_year', 'minimum_nights', 'number_of_reviews', 'review_rate_number', 'calculated_host_listings_count', 'availability_365']
for col in columns:
    copy_data[col] = copy_data[col].astype(int)

copy_data[columns].dtypes

construction_year                 int64
minimum_nights                    int64
number_of_reviews                 int64
review_rate_number                int64
calculated_host_listings_count    int64
availability_365                  int64
dtype: object

In [54]:
copy_data['last_review'] = pd.to_datetime(copy_data['last_review'], errors='coerce')

In [55]:
copy_data['last_review'].min()

Timestamp('2012-07-11 00:00:00')

In [56]:
copy_data.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
lat                                      float64
long                                     float64
instant_bookable                            bool
cancellation_policy                       object
room_type                                 object
construction_year                          int64
price                                    float64
service_fee                              float64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
review_rate_number                         int64
calculated_host_listings_count             int64
availability_365                           int64
house_rules                               object
median_price_by_type_and_group           float64
is_host_identity_verified                   bool
dtype: object

In [57]:
# проверим нет ли у нас дат из будущего
today = pd.to_datetime(datetime.datetime.today().date())
copy_data[copy_data.last_review > today]

id                              name      host_id host_name  \
255  1142173       Beautiful Landmarked Duplex  87944779917     Baker   
318  1176967                           unknown  70084472212    Barnes   
483  1268097  Modern Space in Charming Pre-war  13746585241  Adelaide   

    neighbourhood_group neighbourhood       lat      long  instant_bookable  \
255            Brooklyn    Greenpoint  40.72945 -73.95511              True   
318            Brooklyn    Greenpoint  40.72488 -73.95018              True   
483           Manhattan        Harlem  40.82411 -73.94934             False   

    cancellation_policy  ... minimum_nights  number_of_reviews  last_review  \
255            moderate  ...              3                124   2058-06-16   
318            flexible  ...              2                  1   2026-03-28   
483            flexible  ...              2                 41   2040-06-16   

     reviews_per_month  review_rate_number  calculated_host_listings_count  \
255               1.22                   4                               3   
318               0.01                   3                               1   
483               0.43                   3                               2   

    availability_365                                        house_rules  \
255              230  No rules per say. I just ask that you respect ...   
318               73  Check-In is 3pm.  Check-Out is 12 Noon.  In or...   
483               47                            No house rules provided   

     median_price_by_type_and_group  is_host_identity_verified  
255                           626.0                      False  
318                           624.0                       True  
483                           619.0                       True  

[3 rows x 24 columns]

In [58]:
# удалим строки с отзывами из будущего
print(f'Размер датасета до изменения: {copy_data.shape}')
copy_data = copy_data[(copy_data.last_review <= today)]
print(f'Размер датасета после изменения: {copy_data.shape}')

Размер датасета до изменения: (102546, 24)
Размер датасета после изменения: (86655, 24)


In [59]:
copy_data['minimum_nights'].describe()

count    86655.000000
mean         7.455565
std         27.908866
min       -365.000000
25%          2.000000
50%          3.000000
75%          5.000000
max       5645.000000
Name: minimum_nights, dtype: float64

In [60]:
copy_data[copy_data['minimum_nights'] < 0].shape

(11, 24)

In [61]:
# выведем значения, которые меньше 0
copy_data.loc[copy_data['minimum_nights'] < 0, 'minimum_nights']

176      -10
352       -5
398       -1
421      -10
441      -12
478       -2
525       -3
42500   -365
42538   -200
69749   -125
91357    -10
Name: minimum_nights, dtype: int64

In [62]:
# заменим 1 все значения меньше нуля
copy_data.loc[copy_data['minimum_nights'] < 0, 'minimum_nights'] = 1

In [63]:
copy_data[copy_data['house_rules'] == '#NAME?'].shape

(2108, 24)

In [64]:
# заменим значения '#NAME?' на 'No house rules provided'
copy_data.loc[copy_data['house_rules'] == '#NAME?', 'house_rules'] = 'No house rules provided'

In [65]:
copy_data[copy_data['house_rules'] == '#NAME?'].shape

(0, 24)

In [76]:
copy_data['availability_365'].describe()

count    86655.000000
mean       141.150009
std        131.787733
min          0.000000
25%          6.000000
50%        101.000000
75%        266.000000
max        365.000000
Name: availability_365, dtype: float64

In [70]:
# выведем значения, которые меньше 0
copy_data.loc[copy_data['availability_365'] < 0, 'availability_365']

Series([], Name: availability_365, dtype: int64)

In [69]:
# заменим 1 все значения меньше нуля
copy_data.loc[copy_data['availability_365'] < 0, 'availability_365'] = 1

In [75]:
# выведем значения, которые > 365
copy_data.loc[copy_data['availability_365'] >365, 'availability_365']

Series([], Name: availability_365, dtype: int64)

In [73]:
# заменим 365 все значения
copy_data.loc[copy_data['availability_365'] >365, 'availability_365'] = 365

In [77]:
copy_data.duplicated().sum()

np.int64(480)

In [78]:
# посмотрим на часть дубликатов
copy_data[copy_data.duplicated()].head(3)

id                                            name      host_id  \
102058  35506831  Master Bedroom with private Bathroom & Balcony  55110690425   
102059  35507383              Cozy 2 br in sunny Fort Greene apt  80193772189   
102060  35507935             Duplex w/ Terrace @ Box House Hotel  72991962259   

                  host_name neighbourhood_group neighbourhood       lat  \
102058               UZeyir              Queens       Maspeth  40.74056   
102059                Sally            Brooklyn   Fort Greene  40.68701   
102060  The Box House Hotel            Brooklyn    Greenpoint  40.73756   

            long  instant_bookable cancellation_policy  ... minimum_nights  \
102058 -73.90635              True              strict  ...              1   
102059 -73.97555             False            flexible  ...              3   
102060 -73.95350             False              strict  ...              3   

        number_of_reviews  last_review  reviews_per_month  review_rate_number  \
102058                  1   2021-11-14               0.27                   3   
102059                 38   2021-11-13               0.27                   3   
102060                 10   2021-11-13               0.08                   3   

        calculated_host_listings_count availability_365  \
102058                               1              339   
102059                               1                0   
102060                              30               32   

                    house_rules  median_price_by_type_and_group  \
102058  No house rules provided                           629.0   
102059  No house rules provided                           624.0   
102060  No house rules provided                           836.0   

        is_host_identity_verified  
102058                      False  
102059                       True  
102060                       True  

[3 rows x 24 columns]

In [79]:
copy_data[copy_data.id==35506831]

id                                            name      host_id  \
62476   35506831  Master Bedroom with private Bathroom & Balcony  55110690425   
102058  35506831  Master Bedroom with private Bathroom & Balcony  55110690425   

       host_name neighbourhood_group neighbourhood       lat      long  \
62476     UZeyir              Queens       Maspeth  40.74056 -73.90635   
102058    UZeyir              Queens       Maspeth  40.74056 -73.90635   

        instant_bookable cancellation_policy  ... minimum_nights  \
62476               True              strict  ...              1   
102058              True              strict  ...              1   

        number_of_reviews  last_review  reviews_per_month  review_rate_number  \
62476                   1   2021-11-14               0.27                   3   
102058                  1   2021-11-14               0.27                   3   

        calculated_host_listings_count availability_365  \
62476                                1              339   
102058                               1              339   

                    house_rules  median_price_by_type_and_group  \
62476   No house rules provided                           629.0   
102058  No house rules provided                           629.0   

        is_host_identity_verified  
62476                       False  
102058                      False  

[2 rows x 24 columns]

In [80]:
print(f'Размер датасета до изменения: {copy_data.shape}')
copy_data = copy_data.drop_duplicates()
print(f'Размер датасета после изменения: {copy_data.shape}')

Размер датасета до изменения: (86655, 24)
Размер датасета после изменения: (86175, 24)


# Исследование


In [81]:
neighborhood_prices = copy_data.groupby('neighbourhood')['price'].mean().sort_values(ascending=False).head(10)
print("Топ-10 районов с самыми высокими средними ценами на жилье:")
print(neighborhood_prices)

Топ-10 районов с самыми высокими средними ценами на жилье:
neighbourhood
New Dorp                  1048.000000
Chelsea, Staten Island    1042.000000
Little Neck                906.333333
New Dorp Beach             849.500000
Riverdale                  840.923077
Arden Heights              804.888889
Jamaica Hills              800.250000
Midland Beach              792.666667
Shore Acres                780.214286
Mill Basin                 775.142857
Name: price, dtype: float64


In [82]:
property_price_pivot = copy_data.pivot_table(index='neighbourhood_group', columns='room_type', values='price', aggfunc='mean')
print("\nСредняя цена в зависимости от типа недвижимости и района:")
property_price_pivot


Средняя цена в зависимости от типа недвижимости и района:


room_type            Entire home/apt  Hotel room  Private room  Shared room
neighbourhood_group                                                        
Bronx                     628.290287         NaN    628.070370   600.431818
Brooklyn                  625.965440  711.714286    628.884910   624.851200
Manhattan                 622.873509  681.870000    622.705337   635.721902
Queens                    625.003513  433.250000    634.851509   641.477352
Staten Island             636.364896         NaN    609.348148   695.250000

In [83]:
reviews_per_month_grouped = copy_data.groupby('neighbourhood_group')['reviews_per_month'].mean().sort_values(ascending=False)
print("\nAverage number of reviews per month for each neighborhood group:")
reviews_per_month_grouped


Average number of reviews per month for each neighborhood group:


neighbourhood_group
Queens           1.831982
Bronx            1.749881
Staten Island    1.722352
Brooklyn         1.309417
Manhattan        1.263436
Name: reviews_per_month, dtype: float64

In [84]:
copy_data.groupby(['neighbourhood_group'])['review_rate_number'].mean().sort_values(ascending=False)

neighbourhood_group
Staten Island    3.403073
Bronx            3.320819
Queens           3.320633
Manhattan        3.261535
Brooklyn         3.250552
Name: review_rate_number, dtype: float64

In [85]:
# Group by Construction year and calculate average price and number of reviews
yearly_stats = copy_data.groupby("construction_year").agg({"price": "mean"})
yearly_stats

price
construction_year            
2003               623.544241
2004               633.167610
2005               622.317420
2006               630.982809
2007               627.175316
2008               638.368969
2009               620.631310
2010               622.589331
2011               630.749942
2012               628.772717
2013               619.049697
2014               629.706497
2015               619.223299
2016               625.104774
2017               633.544227
2018               623.828638
2019               611.187328
2020               619.727527
2021               630.930282
2022               630.326928

Выводы
- Три самых дорогих для жилья района - New Dorp, Chelsea, Staten Island и Fort Wadsworth
Масимальная средняя цена у недвижимости типа Shared room в районе Staten Island
- В таких районах как Queens и Staten Island в среднем оставляют больше всего отзывов
- В среднем низкие оценки имеют объявления из районов Brooklyn и Manhattan
- В среднем цена не сильно меняется от года постройки